src/android/ManagerService.java

```diff
@@ -0,0 +1,476 @@
+                }});
+
+                sendCallback(data);
+            }
+        }
+
+        @Override
+        public void onError(final Context context, final UploadInfo uploadInfo, final Throwable exception) {
+            String errorMsg = exception != null ? exception.getMessage() : "unknown exception";
+
+            JSONObject data = new JSONObject(new HashMap() {{
+                put("id", uploadInfo.getUploadId());
+                put("state", "FAILED");
+                put("error", "upload failed: " + errorMsg);
+                put("errorCode", exception instanceof UserCancelledUploadException ? -999 : 0);
+            }});
+
+            deletePendingUploadAndSendEvent(data);
+        }
+
+        @Override
+        public void onSuccess(Context context, UploadInfo uploadInfo, ServerResponse serverResponse) {
+            JSONObject data = new JSONObject(new HashMap() {{
+                put("id", uploadInfo.getUploadId());
+                put("state", "UPLOADED");
+                put("serverResponse", serverResponse.getBodyString());
+                put("statusCode", serverResponse.getCode());
+            }});
+
+            deletePendingUploadAndSendEvent(data);
+            logMessage("onSuccess: " + data);
+        }
+
+        @Override
+        public void onCompleted(Context context, UploadInfo uploadInfo) {
+            stopServiceIfComplete();
+        }
+
+        @Override
+        public void onCompletedWhileNotObserving() {
+            stopServiceIfComplete();
+        }
+    };
+
+    public void sendCallback(JSONObject obj) {
+        if (this.connectedPlugin != null) {
+            this.connectedPlugin.callback(obj);
+        }
+    }
+
+    public void deletePendingUploadAndSendEvent(JSONObject obj) {
+        String id = null;
+
+        try {
+            id = obj.getString("id");
+        } catch (JSONException error) {
+            logMessage(String.format("eventLabel='Uploader could not delete pending upload' error='%s'", error.getMessage()));
+        }
+
+        logMessage(String.format("eventLabel='Uploader delete pending upload' uploadId='%s'", id));
+        PendingUpload.remove(id);
+        createAndSendEvent(obj);
+    }
+
+    public void createAndSendEvent(JSONObject obj) {
+        UploadEvent event = UploadEvent.create(obj);
+        sendCallback(event.dataRepresentation());
+    }
+
+    public void stopServiceIfComplete() {
+        if (PendingUpload.count(PendingUpload.class) == 0) {
+            if (this.connectedPlugin == null) {
+                Intent intent = new Intent(ManagerService.this, ManagerService.class);
+                this.requestObserver.unregister();
+                this.requestObserver = null;
+
+                stopService(intent);
+            } else {
+                Intent intent = new Intent(ManagerService.this, ManagerService.class);
+                stopService(intent);
+
+                startForegroundNotification();
+            }
```

Comment:
Est-ce qu'on peut avoir quelque chose comme ça ?
```suggestion
            if (this.connectedPlugin == null) {
                Intent intent = new Intent(ManagerService.this, ManagerService.class);
                this.requestObserver.unregister();
                this.requestObserver = null;
                stopService(intent);
            }
```
Et pour le `startForegroudNotification` manquant, c'est la ligne
```
UploadServiceConfig.setNotificationHandlerFactory((uploadService) -> new NotificationHandler(uploadService, mainActivity));
```
qui donne les `inputTitle` et `inputContent` comme
```
UploadServiceConfig.setNotificationHandlerFactory((uploadService) -> new NotificationHandler(uploadService, mainActivity, this.inputTitle, this.inputContent));
```
et c'est le `NotificationHandler` qui remet le title et content par défaut une fois terminé avec les uploads ?
(ça peux même être passé comme un hash d'options ?)

Reply:
Le NotificationHandler s'occupe seulement des uploads et la notification n'est invoquée qu'à l'ajout des uploads

Reply:
Le NotificationHandler s'occupe seulement des uploads et la notification n'est invoquée qu'à l'ajout d'images

Reply:
oui mais il ne peut pas savoir à quel moment tous les uploads sont terminés

Reply:
Et quand `this.uploadCount == 0` ? Dans `NotificationHandler`, `onCompleted` mets bien à jour `this.uploadCount`... ?

Reply:
ouais, ça doit passer quand plus d'upload en cours mais le lancement ça va y faire